File: 03-第3周/coding/Stack.py
class Stack:
    def __init__(self):
        self.stack = []
    
    def push(self, x):
        self.stack.append(x)
    
    def pop(self):
        if not self.isEmpty():
            return self.stack.pop()
        else:
            raise IndexError("Stack is empty")
    
    def top(self):
        if not self.isEmpty():
            return self.stack[-1]
        else:
            raise IndexError("Stack is empty")
    
    def isEmpty(self):
        return len(self.stack) == 0

stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)
print(stack.top())  # Output: 3
print(stack.pop())  # Output: 3
print(stack.top())  # Output: 2
print(stack.stack)  # Output: [1, 2]

File: 03-第3周/coding/Stack.cpp
#include <iostream>

class Stack{
    int capacity=5;
    int* arr = new int[capacity];
    int top = -1;
    public:
        Stack(int x){
            arr[0] = x;
            top = 0;
        }
        void push(int x){
            if(top == capacity - 1){
                capacity *= 2;
                arr = (int*)realloc(arr, capacity * sizeof(int));
            }
            arr[++top] = x;
            return;
        }
        void pop(){
            if(top>=0){
                top --;
                return;
            }else{
                std::cout << "stack is empty" << std::endl;
            }
        }
        void print(){
            for(int i = 0; i <= top; i++){
                std::cout << arr[i] << " ";
            }
            std::cout << std::endl;
        }
};

int main(){
    Stack s(1);
    s.push(2);
    s.push(3);
    s.print();
    s.pop();
    s.pop();
    s.pop();
    s.pop();
    s.print();
    return 0;
}